### ISSUE TITLE
Testing MIB model export with bioimage.io

### ISSUE THREAD

**Issue by Ajaxels** [bioimage-io/archive.bioimage.io#324]
Hi,
I have prepared a small test network (attached) that has been trained in DeepMIB and exported in onnx and tensorflow/keras formats.
The network is a simple Unet with two down-sampling steps:
![image](https://user-images.githubusercontent.com/14821531/213246677-00adecf3-f090-49f9-9820-b66c9222bee5.png)

The network was trained to detect white spots over a black backfground:
![image](https://user-images.githubusercontent.com/14821531/213246913-97d56257-bdbf-4d19-80c1-4861fcf70601.png)

The expected results of the network application (the test image in the attached archive):
![image](https://user-images.githubusercontent.com/14821531/213247050-d14f0a68-fea1-470b-8e90-3453c7724d4f.png)

Network settings;
Input patch size: 224x224x1
Padding: same
Encoder depth: 2
Number of classes: 2
Number of first filters: 2
Filter size: 3

There is a readme file for the tensorflow model that describes how it can be converted into TensorFlow SavedModel format.

It would be great to know whether these models can be run on bioimage.io

Thanks!

Best regards,
Ilya
[DeepMIB_test.zip](https://github.com/bioimage-io/bioimage.io/files/10448859/DeepMIB_test.zip)

**Comment by oeway** [comment]
Hi Ilya, thanks for your interest in contributing your models and connect your tool to the model zoo. Could you please take a look at our docs here:  https://bioimage.io/docs/#/contribute_models/README

cc @esgomezm is there other resources or help we can provide to Ilya? Please feel free to comment here.

**Comment by Ajaxels** [comment]
Hi Wei,
we had discussion yesterday on the meeting and the idea was that I provide via issues a small test network exported from Matlab as a quick test to check compatibility with bioimage. Depending on how it goes, I can continue and write a proper output function to generate required metadata.

**Comment by esgomezm** [comment]
@IvanHCenalmor can you please take a look at this one? I would suggest downloading the model and seeing if you can consume it with the `bioimageio.core` library or make the needed changes to make it work. I would then keep track in this issue so @Ajaxels can take a look at it. Does this sound good?

**Comment by IvanHCenalmor** [comment]
Hi Ilya, 
I will take a look at your model and check if it can be used in `bioimage.core`, if anything I will let you know :)

**Comment by Ajaxels** [comment]
@IvanHCenalmor 
awesome Ivan, thank you!

**Comment by constantinpape** [comment]
I had a quick look into this, and conversion of the model into the bioimage.io format works (see snippet for creating a  bioimage.io model from the ONNX weights below). The model seems to run, however it cannot be properly tested because the zip only contains the model input and not the expected model output. (I just used the png for the model output from above, but that does not work).

@Ajaxels: can you please provide the expected output for this model in a suitable format (e.g. tif) as well?


```python
import imageio
import numpy as np
from bioimageio.core.build_spec import build_model

# create the test input and output data as numpy arrays:

# the input -> the model seems to expect inputs of size 224 x 224 pixels, so we crop to this size
# (if not the model test throws an error)
input_ = imageio.imread("../DeepMIB_test/Spots2D_01.tif")[:224, :224]
output = imageio.imread("../DeepMIB_test/test-out.png")
print(input_.shape)
print(output.shape)
np.save("../DeepMIB_test/test-in.npy", input_[None, None])
np.save("../DeepMIB_test/test-out.npy", output[None])

# create doc file
with open("../DeepMIB_test/doc.md", "w") as f:
    f.write("# My Doc")

build_model(
    weight_uri="../DeepMIB_test/onnx/Unet_2layers_224px.onnx",
    test_inputs=["../DeepMIB_test/test-in.npy"],
    test_outputs=["../DeepMIB_test/test-out.npy"],
    input_axes=["bcyx"],
    output_axes=["byxc"],
    output_path="./model-onnx.zip",
    name="test",
    description="test",
    authors=[{"name": "test"}],
    tags=["test"],
    documentation="../DeepMIB_test/doc.md",
    cite=[{"text": "Test et al.", "doi": "test"}],
    opset_version=10,  # IMPORTANT: SET THE CORRECT OPSET VERSION (10 is just a guess)
)
```

**Comment by Ajaxels** [comment]
@constantinpape , fantastic! thanks!
Here is two images for the tests:
[test_image.zip](https://github.com/bioimage-io/bioimage.io/files/11099514/test_image.zip)
The label image (Labels_Spots2D_01.tif) has 0 and 1 indices.

**Comment by constantinpape** [comment]
Thanks @Ajaxels!

The labels are still not quite in the format that we would need; but we don't need to get stuck on this for now.
I managed to produce a valid model export from your weights (using predictions from the model as the expected outputs).

I will bring this up in our meeting today and we will discuss how to proceed and then get back to you.

**Comment by Ajaxels** [comment]
@constantinpape ,
what format you would like to have? should it be 0 and 255?

**Comment by constantinpape** [comment]
What we would need is the exact data that comes out of the model.
Given this pseudocode:
```python
input_ = load_input_patch()  #  the input for the model, which has shape 1x1x224x224 (batch x channel x width x height)
prediction = model(input_)
# prediction has shape 1 x 2 x 224 x 224 (batch x channel x width x height)
```
we would need to have the `prediction` data. (Note that your model only accepts inputs of shape 224 x 224, at least when using the onnx weights that you have provided). In particular this means that the data should have 2 channels and should be floating point (presumably in the range 0 to 1).

But as I said, for now I can just generate the predictions myself and everything seems to work. (This should not be done for actual model exports though, because the goal here is to provide test data for the model to check that the export was done correctly, but for our initial tests this is fine.)

**Comment by Ajaxels** [comment]
you mean probability map (scrores), right? i.e. not the binary segmentation?

**Comment by constantinpape** [comment]
> you mean probability map (scrores), right?

Exactly.

**Comment by Ajaxels** [comment]
I can make that one as well. The problem is to find a suitable format. Can you read mat-files? 
I can probably also convert it to numpy matrix in a similar way as you have in the script above.

**Comment by constantinpape** [comment]
> Can you read mat-files?

mat files are fine (they can be read to numpy by scipy)

**Comment by Ajaxels** [comment]
I scale predictions from floats 0-1 to uint8: 0-255 for the export.
I re-scaled the predictions back to 0-1, but it has probably caused some rounding of values...
[Score_Spots2D_01.zip](https://github.com/bioimage-io/bioimage.io/files/11100343/Score_Spots2D_01.zip)

I can add an option to generate score files in range of 0-1 to make the procedure more straight forward.

**Comment by Ajaxels** [comment]
@constantinpape 
I added an option to generate predictions in the range 0-1 as single class. The resulting MAT-file has 3 variables:
imgOut: [2075×2125×2 single] -> predictions
imgVariable: 'imgOut' -> variable that encodes name of the variable with predictions
options: [1×1 struct] -> structure with additional options, at the moment has a sequence of axes:
options.dimOrder -> 'yxczt'

[Score_Spots2D_01.zip](https://github.com/bioimage-io/bioimage.io/files/11101006/Score_Spots2D_01.zip)

**Comment by constantinpape** [comment]
Thanks @Ajaxels.

The format is correct and now this almost works. There is still some mismatched between the scores you provide in the `.mat` file and the predictions from the onnx model.
There are two potential reasons for this:
- differences at the boundary because of prediction with tiling
- differences in the ONNX version

I will follow up on this later.

**Comment by constantinpape** [comment]
I now created a gist to collect all the steps for the model conversion: https://gist.github.com/constantinpape/de4ad27fad437129867716a790c70740.

@Ajaxels: as I mentioned earlier there are two potential issues that cause the test in there to fail:
- tiling artifacts because prediction is run differently for the full image.
    - Could provide an input image and the corresponding scores / probability output that come from a single pass of the model, without tiling or cropping the results? Note that the ONNX models expects inputs of the exact size 224x224 pixels.
- a difference in the ONNX Opset Version. I have just randomly set this to 10, and the model runs with this version, but a version mismatch may cause some numerical issue.
    - Could you find out the Opset Version of your ONNX export?

**Comment by Ajaxels** [comment]
@constantinpape 
Sorry I only today realized that you added more comments!

Here is the updated version, where I modified the input images and the network to detect larger spots on noisy images:
Input image:
![image](https://user-images.githubusercontent.com/14821531/231746265-1ab9220a-0a0c-456f-a82b-d38129988914.png)
Input image+ground truth:
![image](https://user-images.githubusercontent.com/14821531/231746369-7ef6f135-a7da-48ba-87d8-314a8b3daddc.png)

The tested images have the dimensions matching the input patch size of the network.

Folder in the attached archive:
- onnx folder -> 2 onnx models exported with opset 10
   - DeepMIB_test2\onnx\Unet_2layers_noise_224px_allLayers_OS10.onnx  -> all layers of MATLAB network with the final ""Segmentation-Layer", which is a generalized dice loss.
   - DeepMIB_test2\onnx\Unet_2layers_noise_224px_softmax_OS10.onnx -> the same, but ommiting the segmentation layer, i.e. the final layer is softmax

- prediction_images, images for prediction using the trained network
- prediction_labels, corresponding labels containing the ground truth
- results_labels, resulting bitmap labels (0/1) generated by the network
- results_scores, probabilities for predictions (scale from 0 to 1):
     The resulting MAT-file has 3 variables:
                .imgOut: [2075×2125×2 single] -> predictions
                .imgVariable: 'imgOut' -> variable that encodes name of the variable with predictions
                .options: [1×1 struct] -> structure with additional options, at the moment has a sequence of axes:
                           options.dimOrder -> 'yxczt'

Please let me know if this is enough!
[DeepMIB_test2.zip](https://github.com/bioimage-io/bioimage.io/files/11221718/DeepMIB_test2.zip)

**Comment by constantinpape** [comment]
Hi @Ajaxels,

thanks for providing the new file and it works now!
I have updated the gist that explains how to export the model in python: https://gist.github.com/constantinpape/de4ad27fad437129867716a790c70740
And the exported model is available here: https://owncloud.gwdg.de/index.php/s/VTcGgfUDSydzbSZ

So we can conclude that this works in principle!

**Comment by Ajaxels** [comment]
@constantinpape 
Fantastic! Thanks!
I will try to reproduce this on my side using your instructions!

**Comment by Ajaxels** [comment]
@constantinpape ,
I tested the function and it generates outputs in two places:
- there are files in "exported_model" subfolder:
![image](https://user-images.githubusercontent.com/14821531/232800748-c56a6c1e-9732-443d-85da-137151c904b6.png)
The files are slightly different from the model that you generated.
- in the same directory with the script there is another export:
![image](https://user-images.githubusercontent.com/14821531/232800796-18c93266-c4d2-4cc9-bb70-7e20825e155f.png)
The files look the same as the ones that you've generated, but yaml is missing.

Is it intentional that two models are generated?

**Comment by constantinpape** [comment]
Hi @Ajaxels,
you can ignore all the files generated by the script except for `exported_model/model-onnx.zip`. This contains the zipped model with all relevant files. (The function has to create some temporary files, which is what you see when you run the model. This can be avoided by wrapping this in a temp folder, but I am not sure if this is relevant for you right now.)
